File: server/makc2022--nodejs--nestjs/src/layer1/query/query-extension.ts
/** Copyright (c) 2022 Maxim Kuzmin. All rights reserved. Licensed under the MIT License. */

import { QueryResult } from './query-result';
import { QueryResultWithOutput } from './query-result-with-output';

/** Расширение запроса. */
export class QueryExtension {
  private constructor(
    private self: {
      queryResult?: QueryResult;
      queryResults?: QueryResult[];
    }
  ) {}

  /** Расширить результат запроса.
   * @param queryResult Результат запроса.
   */
  static extQueryResult(queryResult: QueryResult): QueryExtension {
    return new QueryExtension({ queryResult });
  }

  /** Расширить результаты запроса.
   * @param queryResults Результаты запроса.
   */
  static extQueryResults(queryResults: QueryResult[]): QueryExtension {
    return new QueryExtension({ queryResults });
  }

  /** Добавить.
   * @param functionToGetQueryResult Функция для получения результата запроса.
   */
  add(functionToGetQueryResult: () => QueryResult): void {
    const queryResult = functionToGetQueryResult();

    this.self.queryResults.push(queryResult);
  }

  /** Добавить с выходными данными.
   * @template TOutput Тип выходных данных.
   * @param functionToGetQueryResult Функция для получения результата запроса.
   * @param actionToSetOutput Действие по установке выходных данных.
   * @returns Признак успешной установки выходных данных.
   */
  addWithOutput<TOutput>(
    functionToGetQueryResult: () => QueryResultWithOutput<TOutput>,
    actionToSetOutput: (output: TOutput) => void
  ): boolean {
    const queryResult = functionToGetQueryResult();

    this.self.queryResults.push(queryResult);

    if (queryResult.isOk && queryResult.output) {
      actionToSetOutput(queryResult.output);

      return true;
    }

    return false;
  }

  /** Добавить асинхронно.
   * @param functionToGetQueryResult Функция для получения результата запроса.
   */
  async addAsync(functionToGetQueryResult: () => Promise<QueryResult>): Promise<void> {
    const queryResult = await functionToGetQueryResult();

    this.self.queryResults.push(queryResult);
  }

  /** Добавить асинхронно с выходными данными.
   * @template TOutput Тип выходных данных.
   * @param functionToGetQueryResult Функция для получения результата запроса.
   * @param actionToSetOutput Действие по установке выходных данных.
   * @returns Признак успешной установки выходных данных.
   */
  async addWithOutputAsync<TOutput>(
    functionToGetQueryResult: () => Promise<QueryResultWithOutput<TOutput>>,
    actionToSetOutput: (output: TOutput) => void
  ): Promise<boolean> {
    const queryResult = await functionToGetQueryResult();

    this.self.queryResults.push(queryResult);

    if (queryResult.isOk && queryResult.output) {
      actionToSetOutput(queryResult.output);

      return true;
    }

    return false;
  }

  /** */
  /// Загрузить.
  /// </summary>
  /// <param name="queryResult">Результат запроса.</param>
  /// <param name="queryResultsToLoad">Результаты запроса для загрузки.</param>
  load(queryResultsToLoad: QueryResult[]): void {
    let isOk = true;

    queryResultsToLoad.forEach((queryResultToLoad) => {
      isOk = isOk && queryResultToLoad.isOk;

      this.self.queryResult.errorMessages = new Set([
        ...this.self.queryResult.errorMessages,
        ...queryResultToLoad.errorMessages,
      ]);

      this.self.queryResult.successMessages = new Set([
        ...this.self.queryResult.successMessages,
        ...queryResultToLoad.successMessages,
      ]);

      this.self.queryResult.warningMessages = new Set([
        ...this.self.queryResult.warningMessages,
        ...queryResultToLoad.warningMessages,
      ]);
    });

    this.self.queryResult.isOk = isOk;
  }
}

File: server/makc2022--dotnet/Makc2022.Layer1/Query/QueryExtension.cs
// Copyright (c) 2022 Maxim Kuzmin. All rights reserved. Licensed under the MIT License.

namespace Makc2022.Layer1.Query
{
    /// <summary>
    /// Расширение запроса.
    /// </summary>
    public static class QueryExtension
    {
        #region Public methods

        /// <summary>
        /// Добавить.
        /// </summary>
        /// <param name="queryResults">Результаты запроса.</param>
        /// <param name="functionToGetQueryResult">Функция для получения результата запроса.</param>
        public static void Add(ICollection<QueryResult> queryResults, Func<QueryResult> functionToGetQueryResult)
        {
            var queryResult = functionToGetQueryResult.Invoke();

            queryResults.Add(queryResult);
        }

        /// <summary>
        /// Добавить с выходными данными.
        /// </summary>
        /// <typeparam name="TOutput">Тип выходных данных.</typeparam>
        /// <param name="queryResults">Результаты запроса.</param>
        /// <param name="functionToGetQueryResult">Функция для получения результата запроса.</param>
        /// <param name="actionToSetOutput">Действие по установке выходных данных.</param>
        /// <returns>Признак успешной установки выходных данных.</returns>
        public static bool AddWithOutput<TOutput>(
            this ICollection<QueryResult> queryResults,
            Func<QueryResultWithOutput<TOutput>> functionToGetQueryResult,
            Action<TOutput> actionToSetOutput
            )
        {
            var queryResult = functionToGetQueryResult.Invoke();

            queryResults.Add(queryResult);

            if (queryResult.IsOk && queryResult.Output != null)
            {
                actionToSetOutput.Invoke(queryResult.Output);

                return true;
            }

            return false;
        }

        /// <summary>
        /// Добавить.
        /// </summary>
        /// <param name="queryResults">Результаты запроса.</param>
        /// <param name="functionToGetQueryResultTask">
        /// Функция для получения задачи на получение результата запроса.
        /// </param>
        /// <returns>Задача.</returns>
        public static async Task AddAsync(
            ICollection<QueryResult> queryResults,
            Func<Task<QueryResult>> functionToGetQueryResultTask
            )
        {
            var queryResult = await functionToGetQueryResultTask.Invoke();

            queryResults.Add(queryResult);
        }

        /// <summary>
        /// Добавить с выходными данными асинхронно.
        /// </summary>
        /// <typeparam name="TOutput">Тип выходных данных.</typeparam>
        /// <param name="queryResults">Результаты запроса.</param>
        /// <param name="functionToGetQueryResultTask">
        /// Функция для получения задачи на получение результата запроса.
        /// </param>
        /// <param name="actionToSetOutput">Действие по установке выходных данных.</param>
        /// <returns>Задача на получение признака успешной установки выходных данных.</returns>
        public static async Task<bool> AddWithOutputAsync<TOutput>(
            this ICollection<QueryResult> queryResults,
            Func<Task<QueryResultWithOutput<TOutput>>> functionToGetQueryResultTask,
            Action<TOutput> actionToSetOutput
            )
        {
            var queryResult = await functionToGetQueryResultTask.Invoke();

            queryResults.Add(queryResult);

            if (queryResult.IsOk && queryResult.Output != null)
            {
                actionToSetOutput.Invoke(queryResult.Output);

                return true;
            }

            return false;
        }

        /// <summary>
        /// Загрузить.
        /// </summary>
        /// <param name="queryResult">Результат запроса.</param>
        /// <param name="queryResultsToLoad">Результаты запроса для загрузки.</param>
        public static void Load(this QueryResult queryResult, IEnumerable<QueryResult> queryResultsToLoad)
        {
            bool isOk = true;

            foreach (var queryResultToLoad in queryResultsToLoad)
            {
                isOk = isOk && queryResultToLoad.IsOk;

                queryResult.ErrorMessages.UnionWith(queryResultToLoad.ErrorMessages);
                queryResult.SuccessMessages.UnionWith(queryResultToLoad.SuccessMessages);
                queryResult.WarningMessages.UnionWith(queryResultToLoad.WarningMessages);
            }

            queryResult.IsOk = isOk;
        }

        #endregion Public methods
    }
}
